indexers/keyvalue/LevelDBIndexer/tests/test_leveldbindexer.py

```diff
@@ -1,61 +1,195 @@
+    d.buffer = text.encode('utf8')
+    d.weight = weight
+    d.length = length
+    return d
+
+
+def get_documents(ids):
+    documents = [
+        [0, 'cat', 0.1, 3],
+        [1, 'dog', 0.2, 3],
+        [2, 'crow', 0.3, 4],
+        [3, 'pikachu', 0.4, 7],
+        [4, 'magikarp', 0.5, 8]
+    ]
+    id_to_document = {d[0]: d for d in documents}
+    data = [
+        (id_to_document[id][0], id_to_document[id], MessageToJson(create_document(*id_to_document[id])).encode('utf-8'))
+        for id in ids
+    ]
+    return zip(*data)
+
+
+def apply_action(idx, action):
+    method, key_to_id = action
+    with idx:
+        keys = key_to_id.keys()
+        ids = key_to_id.values()
+        ids, documents, documents_bytes = get_documents(ids)
+        if method == 'add':
+            idx.add(keys, documents_bytes)
+        elif method == 'delete':
+            idx.delete(keys)
+        elif method == 'update':
+            idx.update(keys, documents_bytes)
+        else:
+            print(f'method {method} is not implemented')
         idx.touch()
         idx.save()
-        save_abspath = idx.save_abspath
-        index_abspath = idx.index_abspath
-    assert os.path.exists(index_abspath)
-    assert os.path.exists(save_abspath)
 
-    with BaseIndexer.load(save_abspath) as searcher:
-        doc = searcher.query('d2')
-        assert doc.id == 2
-        assert doc.length == 3
 
-    rm_files([save_abspath, index_abspath])
+def apply_actions(save_abspath, index_abspath, actions):
+    for action in actions:
+        idx = BaseIndexer.load(save_abspath)
+        apply_action(idx, action)
+    return save_abspath, index_abspath
 
 
-def test_add_query():
-    indexer = LevelDBIndexer(level='doc', index_filename='leveldb.db')
-    run_test(indexer)
+def validate_positive_results(keys, documents, searcher):
+    for key, query_doc in zip(keys, documents):
+        result_doc = searcher.query(key)
+        assert result_doc.id == str(query_doc[0]) * 16
+        assert result_doc.buffer == query_doc[1].encode('utf8')
+        assert round(result_doc.weight, 5) == query_doc[2]
+        assert result_doc.length == query_doc[3]
 
 
-def test_load_yaml():
+def validate_negative_results(keys, searcher):
+    for key in keys:
+        result_doc = searcher.query(key)
+        assert result_doc is None
+
+
+def validate_results(save_abspath, results, negative_results, _validate_positive_results, _validate_negative_results):
+    with BaseIndexer.load(save_abspath) as searcher:
+        if results:
+            keys, ids = zip(*[[k, v] for k, v in results.items()])
+            _, documents, _ = get_documents(ids)
+            _validate_positive_results(keys, documents, searcher)
+        if negative_results:
+            _validate_negative_results(negative_results, searcher)
+
+
+def get_indexers(tmpdir):
+    # test construction from code
+    indexer_1 = LevelDBIndexer(level='doc', index_filename=Path(tmpdir) / 'leveldb.db')
```

Comment:
can u put better naming, what is the intention? `indexer_constructor` and `indexer_load_config`?

Reply:
I renamed them